Docu for Download Service (#857)
corona-warn-app/cwa-server: docs/DOWNLOAD.md

## Download Runner

The Download Runner triggers the download of Diagnosis Keys (DK) from the EFGS. It then triggers the processing of unprocessed batches and of those which caused errors in previous processing attempts.

## Batch Processing

The FederationBatchProcessor processes batches in sequence and persists all DKs that pass validation.

## Diagnosis Key Validation

Validation constraints enforce that each Key is compliant to the specifications.

### Enumerative Validation Constraints

Enumerative validation constraints are stored as environmental variables and can be consulted in the Download Services `application.yaml` (under `services`.`download`.`validation`).

Name                 | Default Value  | Description
---------------------|----------------|-------------
key-length           | 16             | Exact length of accepted DK Data.
allowed-report-types | CONFIRMED_TEST | Accepted ReportTypes.
max-dsos             | 4000           | Accepted upper bound for Days Since Onset of Symptoms.
min-rolling-period   | 0              | Accepted lower bound for Rolling Period.
max-rolling-period   | 144            | Accepted upper bound for Rolling Period.

> min-trl              | 0              | Accepted lower bound for Transmission Risk Level.

Reviewer: is it 0 or 1 as min-trl?

> max-trl              | 8              | Accepted lower bound for Transmission Risk Level.

Reviewer 2: **upper** bound for Transmission...

### Validation Checks

The checks performed on downloaded DKs are as follows:

* Key Data has correct length.
* Days Since Onset of Symptoms is set and between upper and lower bound.
* ReportType is allowed.
* Starting Interval Number is set and valid.

> * Transmission Risk Level is set and between upper and lower bound.

Reviewer: isn't 'either or'? we reject only if TRL & DSOS are missing, right?

* Rolling Period is set and between upper and lower bound.

Implementation details on validation can be found in: [`ValidFederationKeyFilter.java`](/services/download/src/main/java/app/coronawarn/server/services/download/validation/ValidFederationKeyFilter.java).

### Validation Results and Batch Status

> Keys that fail validation will cause a corresponding log entry, but will not hinder batch processing. The FederationBatchStatus indicates if the the validation and batch processing was successful.

Reviewer 2: remove one 'the'

Processing Status    | Description
---------------------|---------------
PROCESSED            | All DKs passed validation.
PROCESSED_WITH_ERROR | At least one DK failed validation.

> ERROR                | Error non-related to DK validation occured.

Reviewer 3: "occured" should be "occurred"
```suggestion
ERROR                | Error non-related to DK validation occurred.
```

> ERROR_WONT_RETRY     | Error non-related to DK validation occured. Retry of batch processing failed aswell.

Reviewer 3: "occured" should be "occurred" and "aswell" should be "as well"
```suggestion
ERROR_WONT_RETRY     | Error non-related to DK validation occurred. Retry of batch processing failed as well.
```

Enumerative validation constraints are stored as environmental variables and can be consulted in the Download Services `application.yaml` (under `services`.`download`.`validation`).

Name                 | Default Value  | Description
---------------------|----------------|-------------
key-length           | 16             | Exact length of accepted DK Data.
allowed-report-types | CONFIRMED_TEST | Accepted ReportTypes.
max-dsos             | 4000           | Accepted upper bound for Days Since Onset of Symptoms.
min-rolling-period   | 0              | Accepted lower bound for Rolling Period.
max-rolling-period   | 144            | Accepted upper bound for Rolling Period.
min-trl              | 1              | Accepted lower bound for Transmission Risk Level.
max-trl              | 8              | Accepted upper bound for Transmission Risk Level.

### Validation Checks

The checks performed on downloaded DKs are as follows:

* Key Data has correct length.
* ReportType is allowed.
* Starting Interval Number is set and valid.
* Rolling Period is set and between upper and lower bound.
* Days Since Onset of Symptoms (DSOS) is set and between upper and lower bound.

> * Transmission Risk Level (TRL) is set and between upper and lower bound.

Reviewer 4: With proto3, all fields are set automatically to its default

If the validation of DSOS or TRL fails, the corresponding value is derived from the other.

Diagnosis Keys are only rejected if both values are missing or invalid.

Implementation details on validation can be found in: [`ValidFederationKeyFilter.java`](/services/download/src/main/java/app/coronawarn/server/services/download/validation/ValidFederationKeyFilter.java).

### Validation Results and Batch Status

Keys that fail validation will cause a corresponding log entry, but will not hinder batch processing. The FederationBatchStatus indicates if the validation and batch processing was successful.

Processing Status    | Description
---------------------|---------------
PROCESSED            | All DKs passed validation.
PROCESSED_WITH_ERROR | At least one DK failed validation.
ERROR                | Error non-related to DK validation occurred.
ERROR_WONT_RETRY     | Error non-related to DK validation occurred. Retry of batch processing failed as well.

## FederationGatewayDownloadService

Handles the actual downloading and parsing of batches.

## BatchDownloadResponse

Java object representing one downloaded and parsed batch.

## FederationKeyNormalizer

> Helps to derive Transmission Risk Level from Days Since Onset of Symptoms (for backwards compatability).

Reviewer 3: "compatability" should be "compatibility"
```suggestion
Helps to derive Transmission Risk Level from Days Since Onset of Symptoms (for backwards compatibility).
```
